ARROW-6904: [Python] Add support for MapArray

BryanCutler opened the pull request:
This adds support for `MapArray` in Python with conversion from a Python sequence of either dictionaries with "key" and "value" fields or a tuple with 2 elements. Additionally, added the API `MapArray.from_arrays` to build a `MapArray` from individual offset, key, value arrays.

BryanCutler commented:
Made this a draft PR because there are a few things to work out, but feel free to review. Mainly, I think https://issues.apache.org/jira/browse/ARROW-6820 should be resolved first. It was really confusing to keep everything straight since in C++ MapType uses 2 fields for "key" and "value" but the MapArray APIs define everything as "key" and "item". I kept everything as key/item for now.

BryanCutler commented on cpp/src/arrow/array.cc:
```diff
@@ -380,19 +382,82 @@ MapArray::MapArray(const std::shared_ptr<DataType>& type, int64_t length,
-  auto pair_list_data = data->Copy();
-  pair_list_data->type = list(pair_data->type);
-  this->ListArray::SetData(pair_list_data);
-  data_->type = data->type;
```
This was causing a problem. Calling `this->ListArray::SetData(pair_list_data)` would set `list_type_ = pair_list_data->type.get()`, then `pair_list_data->type()` was cleaned up and `list_type_` was no longer valid. Then calling `BaseListArray::value_type()` would crash.

BryanCutler commented on cpp/src/arrow/array/builder_nested.cc:
```diff
@@ -62,31 +66,23 @@ Status MapBuilder::Resize(int64_t capacity) {
   (*out)->type = type();
-  (*out)->child_data[0] =
-      ArrayData::Make((*out)->type->child(0)->type(), keys_data->length, {nullptr},
-                      {keys_data, items_data}, 0, 0);
```
This seemed a bit odd, using a list of keys builder and a separate item builder. I changed it to use use a list builder with a child struct builder. This way I could use this builder to convert a py sequence with the ListConverter as-is.  The only issue was that when using the separate key_builder and item_builders, the struct_builder needs a call to `Append`\`AppendValues` before adding the next index in the MapArray.

BryanCutler commented on cpp/src/arrow/array_list_test.cc:
```diff
@@ -520,7 +520,7 @@ TEST_F(TestMapArray, BuildingStringToInt) {
   ASSERT_OK(actual->Validate());
 
-  ASSERT_ARRAYS_EQUAL(*actual, expected);
+  ASSERT_ARRAYS_EQUAL(expected, *actual);
```
TODO: revert

BryanCutler commented:
Tests look to be passing now, I removed draft status. If we decide to change the Map terminology, I can update this.

BryanCutler commented:
ping @wesm @pitrou @bkietz to take a look when you can, thanks!

pitrou commented on cpp/src/arrow/python/python_to_arrow.cc:
```diff
@@ -806,6 +806,45 @@ class ListConverter
+  }
+
+ protected:
+  Status VerifyLastStructAppended() {
```
Is it useful to do this check at every item? Why not once at the end?

BryanCutler replied:
That was my first thought, but the struct data ends up wrong in the list. I think it's because the struct builder length gets used for the list offsets, so it needs to be called at the end of appending items to each map "slot."

pitrou replied:
I don't get it. The following works currently:
```python
>>> ty = pa.list_(pa.struct([('key', pa.string()), ('item', pa.int8())]))                                                                                                
>>> arr = pa.array([None, [], [('foo', 1), ('bar', 2)], [('xyzzy', 3)]], type=ty)                                                                                        
>>> arr                                                                                                                                                                  
<pyarrow.lib.ListArray object at 0x7fc1bfe962e8>
[
  null,
  -- is_valid: all not null
  -- child 0 type: string
    []
  -- child 1 type: int8
    [],
  -- is_valid: all not null
  -- child 0 type: string
    [
      "foo",
      "bar"
    ]
  -- child 1 type: int8
    [
      1,
      2
    ],
  -- is_valid: all not null
  -- child 0 type: string
    [
      "xyzzy"
    ]
  -- child 1 type: int8
    [
      3
    ]
]
```

Map type conversion is morally the same, so it doesn't seem right that it needs such a special treatment.

BryanCutler replied:
When building a Map array from python, the above works fine with the MapBuilder too. The problem is in C++, when the MapBuilder `key_builder()` and `item_builder()` are used. I will add a C++ test that will hopefully make it clearer.

pitrou replied:
But this method is only doing a check, not changing everything. Why is it required to call it at every element?

BryanCutler replied:
It doesn't just do a check, it adjust the struct_builder length, which is used when the `ListBuilder.Append()` is called (`value_builder` below would be the `struct_builder`):

```C++
Status AppendNextOffset() {
    ARROW_RETURN_NOT_OK(CheckNextOffset());
    const int64_t num_values = value_builder_->length();
    return offsets_builder_.Append(static_cast<offset_type>(num_values));
  }
```

It needs to be called every element in the Map so the offsets of the underlying list are correct. If it's only called once at the end, then the list offsets would say all values are in the last element. Doing this for the test `BuildingStringToInt` would fail with this:


```C++
std::vector<int32_t> offsets = {0, 2, 2, 3, 3};
auto expected_keys = ArrayFromJSON(utf8(), R"(["joe", "mark", "cap"])");
auto expected_values = ArrayFromJSON(int32(), "[0, null, 8]");
```
```diff
/home/bryan/git/arrow/cpp/src/arrow/testing/gtest_util.cc:84: Failure
Failed
@@ -0, +0 @@
-[]
+[{key: "joe", value: 0}, {key: "mark", }]
@@ -2, +2 @@
+[{key: "cap", value: 8}]
@@ -3, +4 @@
-[{key: "joe", value: 0}, {key: "mark", }, {key: "cap", value: 8}]
```
So all the list values end up at the last element because the struct_builder length was 0 when building the offsets at each element.

pitrou replied:
Hmm... can you please explain me where in the following code it does the adjustement?
```c++

    auto struct_builder = checked_cast<StructBuilder*>(BASE::value_converter_->builder());
    auto key_builder = struct_builder->field_builder(0);
    if (key_builder->null_count() > 0) {
      return Status::Invalid("Invalid Map: key field can not contain null values");
    }
    return Status::OK();
```

BryanCutler replied:
omg, sorry. I misunderstood from the get-go and I thought this was a question for another part in the code, let me start again.

Yes, you're right this just checks that the map key value is never null. It could be done once at the end, but I thought it would be better to fail fast rather than process the entire sequence before checking that there is a null key.

The difference between Map and List of Structs is that for Map you could not have the following, which this code would check for:

```python
In [7]: arr = pa.array([None, [(None, 2)], [('foo', 1)], []], type=ty)                                                       

In [8]: arr                                                                                                                  
Out[8]: 
<pyarrow.lib.ListArray object at 0x7fdea79a2e88>
[
  null,
  -- is_valid: all not null
  -- child 0 type: string
    [
      null
    ]
  -- child 1 type: int8
    [
      2
    ],
  -- is_valid: all not null
  -- child 0 type: string
    [
      "foo"
    ]
  -- child 1 type: int8
    [
      1
    ],
  -- is_valid: all not null
  -- child 0 type: string
    []
  -- child 1 type: int8
    []
]
```

pitrou replied:
Well, perhaps at least cache the `key_builder` so that you aren't looking it up every time?

BryanCutler replied:
yeah, good idea. will do

pitrou commented on python/pyarrow/tests/test_array.py:
```diff
@@ -576,6 +576,43 @@ def test_list_from_arrays(list_array_type, list_type_factory):
+    keys = np.arange(5)
+    items = np.arange(4)
+    with pytest.raises(ValueError):
+        pa.MapArray.from_arrays(offsets, keys, items)
```
This is cumulating all kinds of errors (bad offsets, keys and items having different lengths). Ideally you'll check each error separately.

pitrou commented on python/pyarrow/types.pxi:
```diff
@@ -281,6 +281,33 @@ cdef class LargeListType(DataType):
         return pyarrow_wrap_data_type(self.list_type.value_type())
 
 
+cdef class MapType(DataType):
```
Should test this in `test_types.py`.

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -731,6 +731,57 @@ cdef class LargeListValue(ArrayValue):
         return result
 
 
+cdef class MapValue(ArrayValue):
```
Perhaps add a test in `test_scalars.py`.

BryanCutler replied:
done

pitrou commented on cpp/src/arrow/array.h:
```diff
@@ -653,6 +654,25 @@ class ARROW_EXPORT MapArray : public ListArray {
+  /// \param[in] pool MemoryPool in case new offsets array needs to be
+  /// allocated because of null values
+  /// \param[out] out Will have length equal to offsets.length() - 1
+  static Status FromArrays(const std::shared_ptr<Array>& offsets,
```
It would be good to add C++ tests for this as well.

BryanCutler replied:
done

jorisvandenbossche commented on python/pyarrow/scalar.pxi:
```diff
@@ -731,6 +731,57 @@ cdef class LargeListValue(ArrayValue):
+
+    def as_py(self):
+        """
+        Return this value as a Python list of dicts with pairs.
```
This could also be a single dict with multiple key-item pairs ? 

(I am not familiar enough with a "map type" to know if that is a more natural representation, though)

jorisvandenbossche replied:
Ah, I see below in the tests this is actually a list like `[{'key': key1, 'value': value1}, {'key': key2, 'value': value2}, ..]` and not `[{key1: value1}, {key2: value2}, ...]` (as I had in my head)

So with such a structure, it of course needs to be a list as it has duplicate dict keys. But still wondering if something like `{key1: value1, key2, value2}` is not more natural for Python?

wesm replied:
One trouble with this is that in Schema.fbs we say "the application is responsible for ensuring that the keys are hashable and unique". What about returning a list of 2-tuples?

dhirschfeld replied:
Maybe there could be two methods `as_dict` and `as_list`.

`as_list` would return a list of key/value dicts:
```
[{'key': key1, 'value': value1}, {'key': key2, 'value': value2}, ..]
```

whereas `to_dict` would return a single python dictionary:
```
{'key1': value1, 'key2': value2, ...}
```

The `to_dict` method could then raise an exception if the keys weren't hashable or unique.

BryanCutler replied:
> What about returning a list of 2-tuples?

We could do that, but I thought it would be best to return the same as a list of structs for the default.

wesm commented on python/pyarrow/array.pxi:
```diff
@@ -635,7 +635,7 @@ cdef class Array(_PandasConvertible):
         Parameters
         ----------
-        sequence : ndarray, Inded Series
+        sequence : ndarray, Indexed Series
```
This seems odd. Change this to simply `sequence: ndarray, pandas.Series, or array-like`

wesm commented on python/pyarrow/array.pxi:
```diff
@@ -1259,6 +1259,54 @@ cdef class LargeListArray(Array):
+        ----------
+        offsets : Array (int32 type)
+        keys : Array (any type)
+        items : Array (any type)
```
It's more relaxed than that, though -- these can be any sequences / array-like per the implementation

wesm commented on python/pyarrow/array.pxi:
```diff
@@ -1259,6 +1259,54 @@ cdef class LargeListArray(Array):
+    @property
+    def keys(self):
+        cdef CMapArray* arr = <CMapArray*> self.ap
+        return pyarrow_wrap_array(arr.keys())
```
`pyarrow_wrap_array((<CMapArray*> self.ap).keys())` would be okay too

dhirschfeld commented:
I'm quite interested in a pyarrow map type but I'm curious about the api here. Coming from Python it seem a bit awkward to have a map type be represented as a list of key/value dictionaries - i.e. an individual dict for each entry.

Is there some reason a `pa.map` isn't represented directly by a single Python dictionary?

wesm commented:
I think we should add a method to convert `MapValue` to a normal Python dict but having that as the default representation is not appropriate as we cannot guarantee key uniqueness

BryanCutler commented on cpp/src/arrow/array.cc:
```diff
@@ -380,19 +382,82 @@ MapArray::MapArray(const std::shared_ptr<DataType>& type, int64_t length,
+
 void MapArray::SetData(const std::shared_ptr<ArrayData>& data) {
-  ARROW_CHECK_EQ(data->type->id(), Type::MAP);
+  this->ListArray::SetData(data, Type::MAP);
```
`ListArray::SetData` checks that the type id is correct and that there is 1 child. These should be done before the checks below, which is why I moved it up here.

BryanCutler commented on cpp/src/arrow/array.cc:
```diff
@@ -380,19 +382,82 @@ MapArray::MapArray(const std::shared_ptr<DataType>& type, int64_t length,
+        AllocateBuffer(pool, num_offsets * sizeof(offset_type), &clean_offsets));
+
+    // Copy valid bits, zero out the bit for the final offset
+    // XXX why?
```
Yeah, factoring out sounds good.

BryanCutler commented on cpp/src/arrow/array/builder_nested.cc:
```diff
@@ -110,12 +108,25 @@ Status MapBuilder::AppendNull() {
+Status MapBuilder::CheckStructBuilderLength() {
+  // If key/item builders have been appended, adjust struct builder lenght
+  // to match. Struct and key are non-nullable, append all valid values.
+  if (list_builder_->value_builder()->length() < key_builder_->length()) {
```
It's comparing `list_builder_->value_builder()` which is the same as `struct_builder`. I just do the `check_cast` in the `if` statement in case it is not needed. I could move the `checked_cast` to the top though, if it would be more clear?

pitrou replied:
Yes, it would be more clear IMHO. Also, the `checked_cast` is almost a no-op in release mode.

BryanCutler commented on cpp/src/arrow/array/builder_nested.cc:
```diff
@@ -110,12 +108,25 @@ Status MapBuilder::AppendNull() {
 
+Status MapBuilder::CheckStructBuilderLength() {
+  // If key/item builders have been appended, adjust struct builder lenght
+  // to match. Struct and key are non-nullable, append all valid values.
```
If the `key_builder` was used, then the `value_builder` length is not incremented. In this case, we don't want to fail, but adjust the `value_builder` to the correct length.

BryanCutler commented:
>I think we should add a method to convert MapValue to a normal Python dict but having that as the default representation is not appropriate as we cannot guarantee key uniqueness

I agree to not merge dicts as the default. I will see about adding an additional method to produce a single dict.

BryanCutler commented:
Thanks all for the review, I will work on an update.

dhirschfeld commented:
Continuing to bikeshed on a *pythonic* api - I like Wes's suggestion above of `as_py` returning a list of 2-tuples. To me a list of 2-element dicts with the same keys is a bit unnatural and more awkward to work with (and less performant - it forces you to loop in python).

Consider:
```python
In [26]: keys = np.arange(100_000)
    ...: values = np.random.randn(100_000)

In [27]: list_of_tuples = list(zip(keys, values))

In [29]: %timeit dict(list_of_tuples)
20.5 ms ± 219 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [30]: %timeit keys, values = zip(*list_of_tuples)
14.4 ms ± 599 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
vs.
```python
In [28]: list_of_dicts = [{'key': key, 'value': value} for key, value in zip(keys, values)]


In [31]: %timeit dict((d['key'], d['value']) for d in list_of_dicts)
50.3 ms ± 347 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [34]: %%timeit
    ...: keys = []
    ...: values = []
    ...: for d in list_of_dicts:
    ...:     keys.append(d['key'])
    ...:     values.append(d['value'])
    ...:     
40.7 ms ± 543 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

pitrou commented:
>  I like Wes's suggestion above of as_py returning a list of 2-tuples. To me a list of 2-element dicts with the same keys is a bit unnatural and more awkward to work with

Agreed.

dhirschfeld commented:
Also, returning a list of tuples would IMHO remove the need for a method which returned a python `dict` as you could then just do `dict(map_obj.as_py())`.

Maybe there would still be some performance benefit but probably not enough to justify a new method.

BryanCutler commented:
Apologies for the delay, I've updated now and I believe I addressed all comments. `MapValue.as_py()` returns a list of 2-tuples now, but let me know if we should add something else too.

BryanCutler commented on cpp/src/arrow/array_list_test.cc:
```diff
@@ -523,6 +526,78 @@ TEST_F(TestMapArray, BuildingStringToInt) {
+  std::shared_ptr<Array> keys_with_null = offsets3;
+  std::shared_ptr<Array> tmp_items = items->Slice(0, offsets3->length());
+  // ASSERT_RAISES(Invalid,
+  //    MapArray::FromArrays(offsets1, keys_with_null, tmp_items, pool_, &tmp));
```
This fails a `ARROW_CHECK_EQ`, but I'm not sure how to test for that - or maybe there should be an explicit check for keys with null values before it gets to that point?

pitrou replied:
There certainly should be an explicit check, IMHO.

pitrou commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -240,20 +240,38 @@ class ARROW_EXPORT MapBuilder : public ArrayBuilder {
+  /// \brief Get builder to append items
+  ///
+  /// Appending an item with this builder should have been proceeded
+  /// by appending a key with key_builder().
```
Should it be "preceded"?

BryanCutler replied:
oops, yes!

pitrou commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -240,20 +240,38 @@ class ARROW_EXPORT MapBuilder : public ArrayBuilder {
   }
 
+ protected:
+  Status AdjustStructBuilderLength();
```
You could mark this "inline" to encourage the compiler to fold it into its callers.

pitrou commented on cpp/src/arrow/array_list_test.cc:
```diff
@@ -523,6 +526,78 @@ TEST_F(TestMapArray, BuildingStringToInt) {
+  ArrayFromVector<Int8Type, int8_t>(key_values, &keys);
+  ArrayFromVector<Int8Type, int8_t>(items_is_valid, item_values, &items);
+
+  auto map_type = map(int8(), int8());
```
It would be more interesting to use different key/item types, e.g. `map(int8(), int16())`, IMHO.

pitrou commented on python/pyarrow/tests/test_array.py:
```diff
@@ -620,6 +620,56 @@ def test_list_from_arrays(list_array_type, list_type_factory):
+    with pytest.raises(ValueError):
+        pa.MapArray.from_arrays(offsets + [6], keys, items)
+
+    # raise on lenght of keys != items
```
"length"?

pitrou commented on python/pyarrow/tests/test_array.py:
```diff
@@ -620,6 +620,56 @@ def test_list_from_arrays(list_array_type, list_type_factory):
+    # raise on keys with null
+    keys_with_null = list(keys) + [None]
+    with pytest.raises(ValueError):
+        pa.MapArray.from_arrays(offsets, keys_with_null, items)
```
`keys_with_null` and `items` have different lengths here. You can simply use
```
pa.MapArray.from_arrays(offsets, items, keys)
```

BryanCutler replied:
oh right, but `items` doesn't have nulls. I'll fix it.

BryanCutler replied:
fixed

pitrou commented on python/pyarrow/tests/test_array.py:
```diff
@@ -620,6 +620,56 @@ def test_list_from_arrays(list_array_type, list_type_factory):
+
+    # raise on lenght of keys != items
+    with pytest.raises(ValueError):
+        pa.MapArray.from_arrays(offsets, keys, np.concatenate([items, items]))
```
`pa.concat_arrays([items, items])` instead?

BryanCutler replied:
any reason? items is a numpy array, so I think I'd have to convert to pa.Array first

pitrou replied:
Ah, I hadn't noticed. It's ok then.

pitrou commented on python/pyarrow/tests/test_convert_builtin.py:
```diff
@@ -1381,3 +1381,57 @@ def test_decimal_array_with_none_and_nan():
+
+
+def test_map_from_tuples():
+    expected = [[(b'a', 1), (b'b', 2)],
```
Should one of the values be `None`?

BryanCutler replied:
done

pitrou commented on python/pyarrow/tests/test_convert_builtin.py:
```diff
@@ -1381,3 +1381,57 @@ def test_decimal_array_with_none_and_nan():
+
+
+def test_map_from_dicts():
+    data = [[{'key': b'a', 'value': 1}, {'key': b'b', 'value': 2}],
```
Should one of the values be `None`?

BryanCutler replied:
done

BryanCutler commented on cpp/src/arrow/array_list_test.cc:
```diff
@@ -523,6 +526,161 @@ TEST_F(TestMapArray, BuildingStringToInt) {
+  RETURN_NOT_OK(builder.Append());
+  RETURN_NOT_OK(field0_builder->AppendValues({0, 1}));
+  RETURN_NOT_OK(field1_builder->AppendValues({1, -1}, {1, 0}));
+  RETURN_NOT_OK(struct_builder->AppendValues(2, NULLPTR));
```
@pitrou this function builds a ListArray of structs, and here `AdjustStructBuilderLength` doesn't need to do anything. In C++ when using the `MapBuilder::key_builder()` and `MapBuilder::item_builder()` instead, it never calls `struct_builder->AppendValues` like in this line, and therefore the length of the `struct_builder` doesn't change

BryanCutler replied:
If `struct_builder->AppendValues` was only called once at the end, then you would get a List with all struct values in the last slot.

BryanCutler commented on cpp/src/arrow/array_list_test.cc:
```diff
@@ -523,6 +526,161 @@ TEST_F(TestMapArray, BuildingStringToInt) {
+  ASSERT_OK(map_builder.Append());
+  ASSERT_OK(key_builder->AppendValues({0, 1}));
+  ASSERT_OK(item_builder->AppendValues({1, -1}, {1, 0}));
+  ASSERT_OK(map_builder.AppendNull());
```
In this case `AdjustStructBuilderLength` needs to be called after `key_builder` and `item_builder` have appended items to fix the `struct_builder` length, so it is called at the beginning of `map_builder.Append***` before starting the next slot.

jorisvandenbossche commented on python/pyarrow/lib.pxd:
```diff
@@ -209,6 +214,18 @@ cdef class LargeListValue(ArrayValue):
+    cdef getitem(self, int64_t i)
+    cdef int64_t length(self)
+
+
```
Need to add MapArray as well in this file? (a bit below, around line 368)

BryanCutler replied:
Yeah, looks like it's like a forward declaration? I added it but not sure why it's needed..

BryanCutler commented:
@pitrou @jorisvandenbossche please take another look, I think the only remaining item is the discussion at https://github.com/apache/arrow/pull/5774#discussion_r342979217

pitrou commented:
Thank you @BryanCutler !
